0012-unicode-strings.md

```diff
@@ -0,0 +1,377 @@
+instead directing users to fallback to external code for efficient access,
+or to sequential access via the new [ordered quantification features](https://github.com/dafny-lang/rfcs/pull/10) (once they are implemented).
+This would be a much more disruptive breaking change for existing code, however.
+String indexing would not be the first instance in Dafny of a simple, mathematical expression
+that supports verification easily but is inefficient at runtime.
+The proposed lazy decoding approach should provide a good balance between the clean expression of concepts
+and efficient, unsurprising runtime behavior.
+
+## Change `char` to mean an extended grapheme cluster
+
+Rather than just augmenting `char` to represent the more abstract Unicode scalar value concept,
+we could push further and define it to mean an [*extended grapheme cluster*](https://unicode.org/reports/tr29/#Grapheme_Cluster_Boundaries),
+which is the Unicode concept closest to what humans perceive visually as a distinct, atomic "character".
+This would make operations on strings as sequences of values correspond more closely to meaningful concepts on text,
+such as `|s|` actually providing a "character count" that aligns with expectations in a word processor.
+
+Although there are definite benefits to this approach for logic that heavily processes text
+(and is the approach that the [Swift](#swift) language takes, for example),
+the definition of how scalar values are grouped into grapheme clusters
+is much more complicated and locale-specific than the simple mathematical rules of common encodings
+like UTF-8 and UTF-16. This means a higher cost and risk of bugs in adding these concepts
+to the core definition of Dafny, and a much higher likelihood of having to change
+the language definition or implementation to support future versions of Unicode.
+It would also mean `char` can no longer be represented as a single integer value,
+instead needing multiple scalar values in some cases,
+which would be a large a deep change to the encoding of `char` in the Boogie prelude and translator.
+
+# Prior art
+[prior-art]: #prior-art
+
+Unicode has a long and messy history, and the approach to supporting Unicode varies dramatically across different programming languages.
+Here is the current state of several popular programming languages, especially those that Dafny currently supports compiling to,
+or could in the future:
+
+## C#:
+
+`char` is an alias for the `System.Char` value type, which represents a single character as a UTF-16 code unit. 
+`string` is an alias for the `System.String` class, which represents an immutable sequence of `char` values.
+The `System.Text.Rune` struct is provided to represent any Unicode scalar value,
+and its API guarantees that invalid values (e.g. surrogates) will be rejected on construction.
+The method `String.EnumerateRunes()` produces the sequence of runes in a string via an `IEnumerator<Rune>`.
```

Comment:
How does this API behave when given a string that is not a valid UTF-16 string?

Reply:
I've added detail on invalid data in several places now.